### DISCUSSION TITLE
How do I Insert a document only if no document exists with the same Value in a Field?

### DISCUSSION THREAD

**Discussion by sbgmk** [ravendb/ravendb#17335] [Q&A] [answered]
I try to Insert a document only if no document exists with the same Value in a Field. 
I want to insert one new workConf with the name "Test" for example.
The below code functions as long the Calls come one after another. If they are parallel i get multiple documents with the same Name in DB. How can i check if there is already a Doc with this Name and only update or throw some exception in one DB Call with multiple Sessions?

Using C# with this Code at the moment:

` if (Exists<WorkConf>("", nameof(WorkConf.Name), "Test").Result)`
   `return DomainResult.Conflict<WorkConfReference>(string.Format(_localizer["Item_Not_Unique"].Value, dto.Name)); `

 `var workConf = new WorkConf { Name = "Test" };`

 `await _session.StoreAsync(workConf);`
`await _session.SaveChangesAsync();`

Found this similar Discussions but unfortunately dont helped me to solve my Problem.
[how Upsert document](https://github.com/ravendb/ravendb/discussions/14766)
[How do I store a document only if no document exists with the same ID?](https://github.com/ravendb/ravendb/discussions/16675)

**Comment by ayende** [comment]
Depending on how safe you want that to be.

See: https://ayende.com/blog/194405-a/ravendb-5-2-simplifying-atomic-cluster-wide-transactions

If you need that to be always safe at the cluster level (so it wouldn't be able to proceed if a majority of the cluster isn't available).
Similar process can be done using single node transactions, if that isn't a concern, using optimistic concurrency

**Reply by sbgmk** [reply]
Thank you for your fast awsner, but unfortunately it dont solves my problem. Maybe i dont explained my problem well enough.

The upper Code is called from Clients and each call has his own Session. Each call inserts one Document for example with the Name = "Test". 

What i want is that the first call inserts the Document and the other calls dont insert a new Document, if there already exists one. 
If i insert the Document with a fixed ID i would get my wanted result, because the Database allowes only one Document with the defined ID. But i want this behave with some Document Field like the upper Code Example with Name = "Test". 

Here some Example how i want it to Work: 

```
--- Call-1 Session 1 Start Same time as Call 2---
--- Call-2 Session 2 Start Same time as Call 1 ---
var workConf = new WorkConf { Name = "Test" };
```
```
await **_session1**.StoreAsync(workConf);
await **_session1**.SaveChangesAsync();
await **_session2**.StoreAsync(workConf);
```

`await **_session2**.SaveChangesAsync(); --- Call-2 Session 2 End -> Throws some exception because WorkConf with Name "Test" already exists ---`

`--- Call-1 Session 1 End -> Insert WorkConf with Name "Test" Successful ---`

But at the moment i get multiple Documents inserted but only want one Unique Document with the Name="Test". So the database has to check this at the moment it Saves the Changes.
In SQL for example you do this with some unique index for columns, so you get only one of each combination.

**Reply by ayende** [reply, marked as answer]
That is where you use optimistic concurrency and create an additional document with the name reservation, as in the associated blog post.
You'll get `ConcurrenctionException` in this case.

**Reply by sbgmk** [reply]
Ok Thank you for the Awnser again. Now i got it to work with optimistic concurrency and your example in the Blog post.
